src/main/java/org/rocstreaming/roctoolkit/NativeObject.java

```diff
@@ -1,18 +1,53 @@
 package org.rocstreaming.roctoolkit;
 
-class NativeObject {
+/**
+ * A <code>NativeObject</code> represents an underlying native roc object.
+ */
+class NativeObject implements AutoCloseable {
 
-    private long ptr;
+    /**
+     * <code>NativeObject</code> finalizer thread.
+     */
+    private final static NativeObjectCleaner NATIVE_OBJECT_CLEANER = NativeObjectCleaner.getInstance();
 
-    NativeObject() {
-        ptr = 0L;
+    /**
+     *  Reference to {@link NativeObjectPhantomReference}.
+     */
+    private final NativeObjectPhantomReference resource;
+
+    static {
+        RocLibrary.loadLibrary();
+        NATIVE_OBJECT_CLEANER.start();
     }
 
+    /**
+     * Construct a <code>NativeObject</code>.
+     *
+     * @param ptr               native pointer to a roc native object
+     * @param destructor        destructor method for closing <code>NativeObject</code>.
+     */
+    protected NativeObject(long ptr, Destructor destructor) {
+        this.resource = NATIVE_OBJECT_CLEANER.register(this, ptr, destructor);
+    }
```

Comment:
> Then suppose one of the following throws an exception:
> 
> new NativeObjectPhantomReference
> set.add
> new AtomicBoolean(true)
> (If understand correctly, all of those lines can throw)

It could throw only `OutOfMemmoryError` and it's not java way to handle such errors :)